internal/controllers/machineset/machineset_controller_status.go

```diff
@@ -0,0 +1,314 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package machineset
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/ptr"
+
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	v1beta2conditions "sigs.k8s.io/cluster-api/util/conditions/v1beta2"
+	clog "sigs.k8s.io/cluster-api/util/log"
+)
+
+// reconcileV1Beta2Status reconciles Machine's status during the entire lifecycle of the machine.
+// This implies that the code in this function should account for several edge cases e.g. machine being partially provisioned,
+// machine being partially deleted but also for running machines being disrupted e.g. by deleting the node.
+// Additionally, this func should ensure that the conditions managed by this controller are always set in order to
+// comply with the recommendation in the Kubernetes API guidelines.
+// Note: v1beta1 conditions are not managed by this func.
+func (r *Reconciler) reconcileV1Beta2Status(_ context.Context, s *scope) {
+	// Update the following fields in status from the machines list.
+	// - replicas
+	// - v1beta2.readyReplicas
+	// - v1beta2.availableReplicas
+	// Also records the names of machines which are in deletion for more than 30 minutes in s.staleDeletingMachines.
+	setReplicas(s.machineSet, s.machines)
+
+	// Conditions
+
+	// Update the ScalingUp and ScalingDown condition, require the above setReplicas function.
+	setScalingUpCondition(s.machineSet, s.machines, s.bootstrapObjectNotFound, s.infrastructureObjectNotFound)
+	setScalingDownCondition(s.machineSet, s.machines, s.bootstrapObjectNotFound, s.infrastructureObjectNotFound, s.owningMachineDeployment)
+
+	// MachinesReady condition: aggregate the Machine's Ready condition
+	setMachinesReadyCondition(s.machineSet, s.machines)
+
+	// MachinesUpToDate condition: aggregate the Machine's UpToDate condition
+	setMachinesUpToDateCondition(s.machineSet, s.machines)
+
+	// TODO Deleting
+}
+
+func setReplicas(ms *clusterv1.MachineSet, machines []*clusterv1.Machine) {
+	// Return early when machines is nil. It's not possible to calculate replica counters.
+	// Conditions may surface to be unknown in this case.
+	if machines == nil {
```

Comment:
Same as in other places. Let's please have an explicit bool surfacing the fact that we couldn't get the Machines and handle it accordingly